homeassistant/components/duwi/__init__.py

```diff
@@ -0,0 +1,248 @@
+            continue
+
+        # Compare room and device type; if different, add device ID to the list
+        if d1.room_no != d2.room_no or d1.device_sub_type_no != d2.device_sub_type_no:
+            ids.append(dev_id)
+
+    return ids
+
+
+async def cleanup_device_registry(
+    hass: HomeAssistant, entry: DuwiConfigEntry, device_ids: list
+) -> None:
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    device_registry = dr.async_get(hass)
+
+    for dev_id, device_entry in list(device_registry.devices.items()):
+        for item in device_entry.identifiers:
+            if item[0] == DOMAIN and (
+                item[1] not in entry.runtime_data.manager.device_map
+                or dev_id in device_ids
+            ):
+                device_registry.async_remove_device(dev_id)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: DuwiConfigEntry) -> bool:
+    """Unload the Duwi platforms associated with the provided config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(
+        entry, SUPPORTED_PLATFORMS
+    ):
+        house_no = entry.data.get(HOUSE_NO)
+        if house_no in hass.data[DOMAIN].get("existing_house"):
+            hass.data[DOMAIN].get("existing_house").remove(house_no)
+        await entry.runtime_data.manager.unload(True)
+
+    return unload_ok
+
+
+class DeviceListener(SharingDeviceListener):
+    """Device Update Listener."""
+
+    def __init__(self, hass: HomeAssistant, manager: Manager) -> None:
+        """Initialize the DeviceListener."""
+        self.hass = hass
+        self.manager = manager
+
+    def update_scene(self, scene: CustomerScene) -> None:
+        """Notify about scene update."""
+        dispatcher_send(self.hass, f"{DUWI_SCENE_UPDATE}_{scene.scene_no}")
+
+    def update_device(self, device: CustomerDevice) -> None:
+        """Update device status notification."""
+        dispatcher_send(self.hass, f"{DUWI_HA_SIGNAL_UPDATE_ENTITY}_{device.device_no}")
+
+    def add_device(self, device: CustomerDevice) -> None:
+        """Handle a new device being added."""
+        # Ensure the device isn't present stale
+        self.async_remove_device(device.device_no)
+        dispatcher_send(self.hass, DUWI_DISCOVERY_NEW, [device.device_no])
+
+    def remove_device(self, device_no: str) -> None:
+        """Schedule removal of a device."""
+        self.hass.add_job(self.async_remove_device, device_no)
+
+    @callback
+    def async_remove_device(self, device_no: str) -> None:
+        """Remove device from Home Assistant."""
+        device_registry = dr.async_get(self.hass)
+        _LOGGER.info("Removing device_no %s", device_no)
+
+        device_entry = device_registry.async_get_device(
+            identifiers={(DOMAIN, device_no)}
+        )
+
+        if device_entry is not None:
+            device_registry.async_remove_device(device_entry.id)
+
+
+class TokenListener(SharingTokenListener):
```

Comment:
When the user is operating the device and issuing commands, if the cloud service indicates that the token authentication has expired, the system will use the user's refresh_token to refresh the token and retrieve the new token_info. The system will then call the update_token method of the class that implements SharingTokenListener to update the local token_info. If both the token and refresh_token have expired, the user will be prompted to re-authenticate.
My logic is to use the new token_info to update the entry.data with the latest token data when notified, or to prompt the user for re-authentication if both the token and the refresh_token have expired.

Reply:
But why is this a class and not just a simple function?

Reply:
Because there is a logic of identity expiration and identity being blacklisted in identity authentication, the use of interfaces here is for greater scalability. In the future, we may need to implement interfaces for identity tokens that are blacklisted

Reply:
There is also a situation where if the QPS of the user's token is too large, the use of the token will be temporarily restricted, and some method of this interface will be called to remind the user. This will be implemented later